torch_xla/csrc/cross_replica_reduces.cpp

```diff
@@ -210,30 +210,41 @@ AllToAllResult BuildAllToAll(xla::XlaOp input, xla::XlaOp token,
   return {reduce_result, token_handler.GetNewToken(reduce_result)};
 }
 
-AllGatherResult BuildAllGather(xla::XlaOp input, xla::XlaOp token, int64_t dim,
+AllGatherResult BuildAllGather(absl::Span<const xla::XlaOp> inputs,
+                               xla::XlaOp token, int64_t dim,
                                int64_t shard_count,
                                const std::vector<std::vector<int64_t>>& groups,
                                bool pin_layout) {
-  std::vector<xla::ReplicaGroup> reduce_groups = CreateReduceGroups(groups);
-  const xla::Shape& input_shape = ShapeHelper::ShapeOfXlaOp(input);
+  std::vector<xla::ReplicaGroup> cc_groups = CreateReduceGroups(groups);
   TokenHandler token_handler(token);
-  xla::XlaOp all_gather_result;
-  if (pin_layout) {
-    torch::lazy::BackendDevice xla_device = bridge::GetCurrentDevice();
-    xla::Shape reduce_shape = MakeArrayShapeFromDimensions(
-        input_shape.dimensions(), input_shape.dynamic_dimensions(),
-        input_shape.element_type(),
-        static_cast<XlaDeviceType>(xla_device.type()));
-    all_gather_result =
-        xla::AllGather(token_handler.GetInput(input, &input_shape), dim,
-                       shard_count, reduce_groups, /*channel_id=*/absl::nullopt,
-                       /*layout=*/reduce_shape.layout());
-  } else {
-    all_gather_result =
-        xla::AllGather(token_handler.GetInput(input, &input_shape), dim,
-                       shard_count, reduce_groups);
+  // TODO: We use pseudo-tokens ATM, which are real values. This need to be
+  // switched to use the real XLA Token once support has been added to XLA
+  // AllGather().
+  ReduceContext cc_ctx = GetReduceContext(inputs);
+  std::vector<xla::XlaOp> result(inputs.size());
+  for (auto& type_ctx : cc_ctx.contexts) {
+    xla::XlaOp all_gather_result;
+    if (pin_layout) {
+      all_gather_result = xla::AllGather(
+          xla::Tuple(inputs[0].builder(), type_ctx.second.ops), dim,
+          shard_count, cc_groups, /*channel_id=*/absl::nullopt,
+          /*layout=*/
+          MakeReduceShape(type_ctx.second.operand_shapes).layout());
+    } else {
+      all_gather_result =
+          xla::AllGather(xla::Tuple(inputs[0].builder(), type_ctx.second.ops),
+                         dim, shard_count, cc_groups);
+    }
+    if (ShapeHelper::ShapeOfXlaOp(all_gather_result).rank() == 0) {
+      for (size_t i = 0; i < type_ctx.second.indices.size(); ++i) {
+        size_t op_idx = type_ctx.second.indices[i];
+        result[op_idx] = xla::GetTupleElement(all_gather_result, i);
+      }
+    } else {
+      result[0] = all_gather_result;
+    }
```

Comment:
I don't see how token is being used here, in all_reduce, we manually append it at the end of `ops`
```
  for (auto& type_ctx : redux.contexts) {
    xla::XlaOp token_op = MaybeConvertTo(chained_token, type_ctx.first);
    type_ctx.second.ops.push_back(token_op);
    type_ctx.second.operand_shapes.push_back(
        ShapeHelper::ShapeOfXlaOp(token_op));
```

Reply:
I saw the `GetOperandList` below, but I think this does not gurante when you have multiple types, each types has a token.

Reply:
Fix the usage of token. 

For multiple types case, I think for now we should ensure same type in the list.